Studio: Add dynamic code block to execute wp-cli commands UI AI assistant

katinthehatsite opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/7526

## Proposed Changes

This PR adds that dynamic code block that allows to execute CLI commands inside that block. It also does the following:

* if the code block contains more than 1 CLI command, the Run button is disabled
* the `Run` button is disabled for code blocks that don't have CLI commands
* when you execute the CLI command, see the output, navigate to another tab and come back, you can see the results of the output preserved 

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Pull the changes from this branch locally
* Start the app with `STUDIO_AI=true npm start`
* Navigate to the Assitant tab
* Ask the assistant to give you a CLI command to list users
* Observe that it produces a code block
* Run the command with `Run` button 
* Confirm that you see the output being rendered
* Navigate to a different tab
* Come back to the AI Assistant
* Observe that the output of the CLI command is preserved:

<img width="872" alt="Capture d’écran, le 2024-06-13 à 10 26 19" src="https://github.com/Automattic/studio/assets/25575134/e47d7aea-3299-4a82-9754-07948eda75c9">

* Try asking for different CLI commands and checking the output

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

Commit:
Enable Run cli button

Commit:
Fix timer for the time of execution

Commit:
Move state to code block to avoid the run button running on all blocks

Commit:
Display output from cli for error and success cases

Commit:
Hide the button for code blocks that have no CLI commands

Commit:
Ensure only code blocks with one wp command execute

Commit:
Add logic for updating messages in useAssistant

Commit:
Add logic for storing the messages

Commit:
Fix type on updateMessage

Commit:
Handle undefined ID

Commit:
Fix types on authenticated view

Commit:
Fix some more types

Commit:
Fix up another warning

Commit:
Cleanup

Commit:
Update trunk changes

Commit:
Merge trunk

Commit:
Split content tab assistant and chat message

Commit:
Fix weird re-render

Commit:
Fix unit tests

katinthehatsite commented:
A note that I noticed that sometimes CLI output could be strange - for instance, I asked the Chat Assistant to give me CLI commands and it suggested the following but running that command gave strange output that requires action:

<img width="868" alt="Capture d’écran, le 2024-06-12 à 21 49 47" src="https://github.com/Automattic/studio/assets/25575134/421a4a15-7744-4b1f-8167-175ebb8f64c7">

We might need to add a better validation of the commands that we want to run to make sure it does not result into the actionable output like this.

Commit:
Strip ANSI codes from errors

Commit:
Move the loggic of processing cli output into a hook

Commit:
Ensure that the DOM updates for the scroll to work

katinthehatsite commented on package.json:
```diff
@@ -115,6 +115,7 @@
 		"hpagent": "1.2.0",
 		"react-markdown": "^9.0.1",
 		"semver": "^7.6.0",
+		"strip-ansi": "^7.1.0",
```
This was necessary to strip ANSI that comes in the output from CLI - otherwise, the Error and Success messages contain ANSI characters and manually removing them is quite tedious as it is hard to match all of them. I felt like this was appropriate to use

kozer commented on src/hooks/use-execute-cli.ts:
```diff
@@ -0,0 +1,71 @@
+		}, 2300 );
+	}, [ content, id, projectPath, updateMessage ] );
+
+	return {
```
[Suggestion]: I find it a bit repetitive. As we call it like this:
```
	const {
			cliOutput,
			cliStatus,
			cliTime,
			isRunning,
			handleExecute,
			setCliOutput,
			setCliStatus,
			setCliTime,
		} = useExecuteWPCLI( content, projectPath, updateMessage, id );

```
it is clear enough that this comes from cli.
I suggest to rename them to:
```
output,
status,
time
setOutput,
setStatus,
setTime
```

kozer commented on src/components/chat-messages.tsx:
```diff
@@ -0,0 +1,202 @@
+			if ( blocks ) {
+				const block = blocks?.find( ( block ) => block.codeBlockContent === content );
+				if ( block ) {
+					setCliOutput( block?.cliOutput ? stripAnsi( block.cliOutput ) : null );
```
[Suggestion] I suggest to change that to `setBlock`, as this is what those variables actually represent

kozer commented:
@katinthehatsite  nice work!
I left some comments/suggestions.
Other than that it works as expected
![image](https://github.com/Automattic/studio/assets/497103/2255cf59-bb83-4ae9-be8e-67ad7a3a994a)

Note that there are some errors in `buildkite/unit-tests` that are related to the new library you introduced.

katinthehatsite commented:
Something that still needs to be looked into but documenting it here for the reviewers:

* when you use Run button and the output appears, the chat scrolls to the bottom of the page. It is because of: https://github.com/Automattic/studio/blob/f756e2aa0a6b7644130d7fa71e3a6b5d204f3790/src/components/content-tab-assistant.tsx#L41 that scrolls to the bottom every time an array of messages updates. It would need to be somehow adjusted to scroll through when the messages are added vs. not scroll through when the output is produced (this happens with updateMessage)
* another thing that the `Run` button does not go to `Run again` once the output is generated. This needs to be adjusted as well

katinthehatsite commented:
Looking at https://github.com/Automattic/dotcom-forge/issues/7526#issuecomment-2165266623 , let's wait until Monday when we decide if we want to keep the `Run` button at all. If not, then we can remove it and otherwise, this PR is ready.

If we decide to keep it, let's fix the two things mentioned above in https://github.com/Automattic/studio/pull/232#issuecomment-2165073454

Commit:
Fix scroll for messages

sejas commented:
@katinthehatsite , would be possible to add an horizontal scroll to the code block?

<img width="661" alt="Screenshot 2024-06-17 at 10 28 25" src="https://github.com/Automattic/studio/assets/779993/fde8170f-19b7-4caf-b0a7-6358f4aa0e89">

Commit:
Fix jest configuration

Commit:
Fix the jest setup

sejas commented:
Let's also avoid displaying the Run button if the wp command has some placeholder `<` `>`.

Commit:
Fix failing unit test

katinthehatsite commented on src/hooks/use-assistant.ts:
```diff
@@ -1,9 +1,16 @@
 
-export interface Message {
+export type Message = {
+	id?: number;
```
I am not 100% sure if this should stay optional or must be mandatory.

katinthehatsite committed:
Merge branch 'trunk' into add/integrate-cli-button-to-execute-in-code-block

Commit:
Fix linter on use-assistant

Commit:
Fix linter on content-tab-assistant

Commit:
Refactor the copy button in code block

Commit:
Refactor button logic for Run and Run again

Commit:
Fix styling on run button

Commit:
Fix buttons

Commit:
Final touches on button styles

Commit:
Hide commands that include placeholders

katinthehatsite commented:
Hi @sejas, since our conversation earlier this morning, I have done the following:

* Refactored the `Run` and `Run again` buttons, reused some of the components that we already have and removed the redundant logic;
* Fixed the failing unit test;
* Resolved conflicts with trunk and all the latest changes how now been integrated with trunk
* Added a check that looks for `<` and `>` characters and disables the `Run` button for these commands

The only other thing I see is that it seems that E2E tests are timing out on Windows 🤔 

For the horizontal scroll, I suggest we open a separate issue and handle it in the follow-up. I am not seeing an obvious fix right away and I would prefer to merge the PR asap as it is starting to grow quite large. What do you think?

Let me know what you think of the changes!

Commit:
Resolve conflict

sejas committed:
Refactor component name to ChatMessage

sejas committed:
Merge branch 'trunk' of github.com:Automattic/studio into add/integrate-cli-button-to-execute-in-code-block

sejas committed:
Fix args passed to executeWPCliInline to be one string

sejas committed:
Use meesage id in useExecuteWPCLI

sejas committed:
Add empty line on package.json

sejas commented on src/components/chat-message.tsx:
```diff
@@ -36,69 +55,47 @@ const InlineCLI = ( { output, status, time }: InlineCLIProps ) => (
+		const containsWPCommand = /\bwp\s/.test( content );
+		const wpCommandCount = ( content.match( /\bwp\s/g ) || [] ).length;
+		const containsSingleWPCommand = wpCommandCount === 1;
+		const containsAngleBrackets = /<.*>/.test( content );
```
We can add more conditions to make this code block only available in a few commands.

fluiddot replied:
I noticed that the AI also uses braces `{}` for argument placeholders. I propose including this symbol in the regular expression. WDYT?

<img width="542" alt="Screenshot 2024-06-26 at 11 50 32" src="https://github.com/Automattic/studio/assets/14905380/1dc3cfab-d921-4d1b-9438-29755650fd40">

fluiddot replied:
I created [a PR](https://github.com/Automattic/studio/pull/309) with improvements in this area.

dcalhoun replied:
Noting that I have also seen the Assistant exclude wrapping characters altogether for placeholder content, which complicates things a good bit. I suppose the only solution for this case would be adjusting the Assistant prompt in hopes that it avoids this approach or searching for `example` terms (which is undoubtedly brittle). 

<img width="543" alt="Screenshot 2024-06-26 at 09 10 58" src="https://github.com/Automattic/studio/assets/438664/41e4ab8a-196d-4b47-8b1c-13db9b621b3e">

fluiddot replied:
Oh, I haven't encountered that but yes, that definitely complicates the detection of placeholders. As you pointed out, seems the best approach would be to tweak the prompt to ensure it uses the same placeholder to avoid cases like this one.

sejas reviewed (approved):
Works as expected.
It keeps the result after changing sites and tabs. 

@katinthehatsite , great work!

https://github.com/Automattic/studio/assets/779993/d9d77319-8113-4c3d-8873-7d8027394ce1

dcalhoun commented on src/components/copy-text-button.tsx:
```diff
@@ -34,6 +40,7 @@ export function CopyTextButton( {
 		}
 		setTimeoutId( setTimeout( () => setShowCopied( false ), timeoutConfirmation ) );
 	}, [ text, timeoutConfirmation, timeoutId ] );
+	console.log( 'Icon Size:', iconSize );
```
I suggest we remove this console log. I presume its addition was oversight.

dcalhoun commented on src/components/content-tab-assistant.tsx:
```diff
@@ -26,23 +26,55 @@ const AuthenticatedView = memo(
-	} ) => (
-		<>
-			{ messages.map( ( message, index ) => (
-				<ChatMessage key={ index } id={ `message-${ index }` } isUser={ message.role === 'user' }>
```
It appears the `id` attribute was removed. The removal introduces a regression where screen readers are no longer able to navigate each message as a group, which was enabled in #248. 

I presume this was oversight and likely originates from a merge conflict resolution.

katinthehatsite replied:
Thanks for catching these and adding the PR to fix. Yeah that was a large conflict merge

dcalhoun reviewed (commented):
I noted a few issues after pulling these changes. I share them here merely to raise awareness. I'll open a PR addressing them.
